ag-master/main.py

```diff
@@ -0,0 +1,447 @@
+            timeout=1,
+        )
+    except requests.exceptions.ReadTimeout:
+        pass
+
+    return dict(jobs=jobs)
+
+
+@app.route("/trigger_jobs", methods=["POST"])
+@check_master
+def trigger_jobs():
+    data = request.json
+    assignment_id = data["assignment_id"]
+    subms = data["subms"]
+    jobs = data["jobs"]
+
+    assignment = Assignment.query.filter_by(ag_key=assignment_id).first()
+    if not assignment:
+        abort(404, "Unknown Assignment")
+
+    subm_batches = [subms[i : i + 100] for i in range(0, len(subms), 100)]
+    job_batches = [jobs[i : i + 100] for i in range(0, len(jobs), 100)]
+
+    for subm_batch, job_batch in zip(subm_batches, job_batches):
+        trigger_job_batch(assignment, subm_batch, job_batch)
+    return dict(success=True)
+
+
+def trigger_job_batch(assignment, ids, jobs):
+    try:
+        requests.post(
+            f"{WORKER_URL}/batch_grade",
+            json=dict(
+                assignment_id=assignment.ag_key,
+                assignment_name=assignment.name,
+                command=assignment.command,
+                jobs=jobs,
+                backups=ids,
+                course_key=assignment.course,
+            ),
+            headers=dict(Authorization=get_secret(secret_name="AG_WORKER_SECRET")),
+            timeout=1,
+        )
+    except requests.exceptions.ReadTimeout:
+        pass
+    return jobs
+
+
+@app.route("/results/<job_id>", methods=["GET"])
+def get_results_for(job_id):
+    job = Job.query.filter_by(job_key=job_id).first()
+    if job and job.status in ("finished", "failed"):
+        return job.status, 200
+    return "Nope!", 202
+
+
+@app.route("/results", methods=["POST"])
+def get_results():
+    job_ids = request.json
+    jobs = Job.query.filter(Job.job_key.in_(job_ids)).all()
+
+    res = {job.job_key: dict(status=job.status, result=job.result) for job in jobs}
+    for job in job_ids:
+        if job not in res:
+            res[job] = None
+
+    return res
+
+
+########################################
+## AG Worker Endpoints
+########################################
+
+
+@app.route("/get_zip", methods=["POST"])
+@check_secret
+def get_zip(course):
+    assignment = Assignment.query.filter_by(
+        name=request.json["name"], course=course.secret
+    ).first()
+    if assignment:
+        bucket = storage.Client().get_bucket(BUCKET)
+        blob = bucket.blob(f"zips/{course.name}-{course.semester}/{assignment.file}")
+        with tempfile.NamedTemporaryFile() as temp:
+            blob.download_to_filename(temp.name)
+            return send_file(temp.name)
+    abort(404)
+
+
+@app.route("/get_submission", methods=["POST"])
+@check_secret
+def get_submission(course):
+    data = request.json
+    id = data["id"]
+    job = Job.query.filter_by(job_key=data["job_id"]).first()
+    assignment = Assignment.query.filter_by(
+        ag_key=job.assignment, course=course.secret
+    )  # validates secret
+    if job and assignment:
```

Comment:
does not validate! You need to check that the query has an output - queries on their own are always truthy, I think

Reply:
though technically the job_id, being secret, is enough